pkg/executor/aggregate/agg_hash_final_worker.go

```diff
@@ -174,19 +223,62 @@ func (w *HashAggFinalWorker) receiveFinalResultHolder() (*chunk.Chunk, bool) {
 	}
 }
 
-func (w *HashAggFinalWorker) run(ctx sessionctx.Context, waitGroup *sync.WaitGroup) {
+func (w *HashAggFinalWorker) run(ctx sessionctx.Context, waitGroup *sync.WaitGroup, partialWorkerWaiter *sync.WaitGroup) {
 	start := time.Now()
-	defer func() {
-		if r := recover(); r != nil {
-			recoveryHashAgg(w.outputCh, r)
+	defer w.cleanup(start, waitGroup)
+
+	partialWorkerWaiter.Wait()
+
+	intestBeforeFinalWorkerStart()
+
+	if w.spillHelper.isSpilledChunksIOEmpty() {
+		err := w.consumeIntermData(ctx)
+		if err != nil {
+			w.outputCh <- &AfFinalResult{err: err}
+			return
 		}
-		if w.stats != nil {
-			w.stats.WorkerTime += int64(time.Since(start))
+	} else {
+		if w.spillHelper.checkError() {
+			return
 		}
-		waitGroup.Done()
-	}()
-	if err := w.consumeIntermData(ctx); err != nil {
-		w.outputCh <- &AfFinalResult{err: err}
 	}
-	w.loadFinalResult(ctx)
+	w.sendFinalResult(ctx)
+}
+
+func (w *HashAggFinalWorker) cleanup(start time.Time, waitGroup *sync.WaitGroup) {
+	if r := recover(); r != nil {
+		recoveryHashAgg(w.outputCh, r)
+	}
+	updateWorkerTime(w.stats, start)
+	waitGroup.Done()
+}
+
+func intestBeforeFinalWorkerStart() {
+	failpoint.Inject("enableAggSpillIntest", func(val failpoint.Value) {
+		if val.(bool) {
+			num := rand.Intn(50)
+			if num == 0 {
```

Comment:
0 <= num < 15, panic
15 <= num < 30, sleep

Reply:
> 0 <= num < 15, panic 15 <= num < 30, sleep

I think `0 <= num < 3, panic, <= 3 num < 6, sleep` is better, as we have 5 final workers in the test.